### PR TITLE
Code Intelligence: Update configuration policy UI

### PR DESCRIPTION
**Pull request by umpox**
**Updates the code intel configuration policy with various changes and improvements.**

Closes https://github.com/sourcegraph/sourcegraph/issues/47111

Related [designs](https://www.figma.com/file/PqRnWEXK9pJPy6T1oxw4nQ/AutoIndex-Configuration-QA?node-id=0%3A1)

### What this PR does do

This PR has quite a lot of small changes that don't make sense to split into different PRs but equally are hard to infer from the diff in a single PR. Here is my log of what has changed (aside from just general fixes and improvements):

- Configuration policy list page
    - [X] Added Button dropdown to replace 3 horizontal buttons.
        - Defaults to HEAD
        - This is similar to Batch Changes. I think this component is a good candidate for Wildcard (will move in future PR)
    - [X] Shifted protected icon to the right, to replace the empty space where the delete icon was and make the list more consistent
        - Note: In a future PR I want to move the default policies into their own space. Haven't done it here as it makes sense to update the BE to allow filtering.
    - [X] Converted policy names into links, so it’s more obvious to the user that they are actionable (and stuff like the default ones are viewable).
- Configuration policy page
    - General
        - [X] Updated input width to be smaller, to make it more obvious this is a form, as per designs.
        - [X] Updated headings to use correct labels and link them with input fields for accessibility
        - [X] Move indexing/retention toggles to left side to match other inputs, as per design.
        - [X] Defaulted the policy name instead of a placeholder, so a user isn’t presented with an error when they first load.
        - [X] Various other wording changes and UX polish
        - [X] Fixed bug where the policy did not validate that the global auto indexing setting was not violated. (It let the user save the policy). Also moved the alert directly below the Save button for easier reading.
    - Configure Git objects
        - [X] Moved the matching results table to be directly below input, for easier reading.
        - [X] Added ability to fetch all matching git objects (no longer limited to just 15)
        - [X] Added scrolling behaviour if user fetches more than 15 objects.
        - [X] Moved loading spinner UX to be contained within input.
        - [X] Made wording clearer when combined with auto-indexing logic
        - [X] Object links now open in a new tab, to avoid users accidentally losing config
    - Add repo patterns
        - [X] Updated UI to match design
        - [X] Fixed input and actions size to make it easier to quickly add/remove filters
        - [X] Improved autofocus logic
        - [X] Changed default “This matches all repositories” text to be clearer with an obvious CTA to add repositories filter.
        - [X] Cleaned up loading spinner and error UI screens.
        - [X] Fixed race condition bug between network and user input due to render logic.
        - [X] Added placeholder to help users understand different patterns
    - Repo list
        - [X] Added ability to fetch all matching repositories (no longer limited to just 15)
        - [X] Added scrolling behaviour if user fetches more than 15 repos.
        - [X] Updated “Too many matching repositories” wording to match design and link directly to site admin settings.
        - [X] Updated repository name logic so that the code host *is* displayed. This matches what would be expected in the filtering logic so it should be a lot easier for the user to understand how to manipulate this list.
        - [X] Repo links now open in a new tab, to avoid users accidentally losing config

### What this PR does not do
- [ ] Shift components around into different structures. I've avoided this mostly, aside from changes that can are required to fix render bugs. This to keep the diff relatively clean. In a future PR I can move this around
- [ ] Significantly change the render logic. There's some improvements we can make to streamline these components a bit. I fixed a few bugs where we are over rendering (and causing UX issues), but avoided minor issues. Time is limited so I'm focusing on areas where we get the biggest UX wins right now.

## Some screenshots

image

image

image

## Test plan

Tested locally

## App preview:

- [Web](https://sg-web-tr-index-configuration-ui.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | -0.00% (-0.12 kb) | -0.00% (-0.12 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/tr/index-configuration-ui/compare-report.html;tab=live_object) for a full comparison between the commits 7f7bbf9 and 653fe90 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by umpox** [comment]
@efritz @rrhyne If you get chance, please take a look and LMK your thoughts on the UX. Keen to iterate on this as required

File: client/web/src/enterprise/codeintel/configuration/shared.ts [hunk 1]
```diff
@@ -15,3 +17,18 @@ export const formatDurationValue = (value: number): string => {
...
+    // If there are no repo patterns, it is assumed that the policy targets all repos.
+    const repoPatterns = policy.repositoryPatterns || []
+
+    return (
+        // User has enabled auto indexing for a policy
+        policy.indexingEnabled &&
+        // Policy isn't targeted at a specific repository
+        !policy.repository &&
+        // Policy does not have a targeted repository pattern.
+        // TODO(#47432): This is flaky as repoPatterns can match all repositories (e.g. '*'). We should return a flag that indicates if this has happened.
```
Comment (efritz) [inline client/web/src/enterprise/codeintel/configuration/shared.ts:31]: 👍 

plan to address this issue before EoW

**Review by efritz — APPROVED** [review]

**Comment by efritz** [comment]
Just to chat about future directions:

> Note: In a future PR I want to move the default policies into their own space. Haven't done it here as it makes sense to update the BE to allow filtering.

Note that the "protected" policies are hard-coded into the instance, so there will only ever be a handful (right now three, and there's no pressure to add another default). Is that your looking for here just a "protected/non-protected" filter on the GraphQL endpoint to fetch all related policies?

>  Added ability to fetch all matching repositories (no longer limited to just 15)

If we ask for all bracnhes on a large monorepo (matching '*' for instance) do we expect to have some text like "See all 12384791874 tags" and expect to render them all? Do we want to add some kind of upper bound here on either the frontend (if over some threshold render a non-actionable link) or backend (maximum allowable return set size)?

**Comment by umpox** [comment]
@efritz 

> Note that the "protected" policies are hard-coded into the instance, so there will only ever be a handful (right now three, and there's no pressure to add another default). Is that your looking for here just a "protected/non-protected" filter on the GraphQL endpoint to fetch all related policies?

That's right. There isn't a nice way to filter these out on the client as it requires fetching the entire list (we also currently paginate and allow filtering, so gets weird). It'd be ideal if we had a simple GQL variable to filter against protected/non-protected policies.

> If we ask for all bracnhes on a large monorepo (matching '*' for instance) do we expect to have some text like "See all 12384791874 tags" and expect to render them all? Do we want to add some kind of upper bound here on either the frontend (if over some threshold render a non-actionable link) or backend (maximum allowable return set size)?

Good point. I have set 1000 as a limit for both repositories and git branches/tags. That feels reasonable enough to me as something which should cover most use cases without leading to users accidentally triggering a huge query.

**Pull request merged**